Basic Mob Spiders I: Ability refactor

Jacquerel opened the pull request:
## About The Pull Request

This morning a man presented me with a mission
![image](https://user-images.githubusercontent.com/7483112/216712023-788f2f35-feaa-4152-acee-a2a58e1e6d2c.png)
I realised that I can achieve this in merely _three_ pull requests.
...because I'm leaving one that adds new AI until after feature freeze february.

This one pulls all of the ability code out of the giant spider mob file and spruces it up a bit in preparation for making them into basic mobs.
It removes an ancient "todo" about merging the spider and genetics web abilities by, doing that.
Spider abilities no longer extend from a shared and reasonably pointless "innate" root ability and also don't require their owners to be spiders to work properly any more.

Also while I was in there I fixed a bug where the "set directive" ability did not actually pass that into the antag datum of a new spider unless they logged out and back in.

## Why It's Good For The Game

Code should smell nicer now.
Removes an essentially duplicated ability.
Clears up some chat log spam which didn't need to be in there.
You can actually set an objective for spiders you create as a spider brood mother.
I guess you can make spiders-man as an admin by giving spider abilities to other mobs.

## Changelog

:cl:
refactor: Tidied up spider ability code, they should all work the same, but please report if they do not.
refactor: Spiders and Geneticists now use the same ability to make webs rather than two virtually-identical ones, which should also perform just as it did before this change.
fix: Setting a directive as a Spider Midwife should properly pass and announce itself to new spiders.
/:cl:

Jacquerel committed:
Refactors spider abilities out into their own files
Moves spider do_after define into defines file
Abilities don't rely on you being a spider any more
Unified spider/genetics web abilities
Fixed directive passing

Jacquerel committed:
Removes outdated comment
Simplifies some web placing logic

Jacquerel commented on code/modules/mob/living/basic/space_fauna/giant_spider/spider_abilities/wrap.dm:
```diff
@@ -0,0 +1,120 @@
+		span_notice("You begin wrapping [to_wrap] into a cocoon."),
+	)
+
+	var/mob/living/simple_animal/animal_owner = owner
+	if(istype(animal_owner))
+		animal_owner.stop_automated_movement = TRUE
```
I'll be honest I mostly kept it because it was existing behaviour, none of these mobs have AI which actually uses these abilities so I am not sure it would ever trigger automated movement while using them anyway except if their AI fails to turn off when they gain a ckey (which to be fair... does happen sometimes, at least with the simple mob version we still have)

Jacquerel committed:
Balloon alert is more descriptive, comes from web.
Simplified web obstruction check.
Removes automated movement pausing for spider actions, given that they're never used by AI in the first place and this system won't work when they do.
Plus it's not like I'm going to make AI for midwives.

Jacquerel commented on code/game/objects/effects/spiderwebs.dm:
```diff
@@ -67,7 +68,7 @@
 			return TRUE
 		if(prob(50))
-			to_chat(mover, span_danger("You get stuck in \the [src] for a moment."))
+			mover.balloon_alert(mover, "stuck!")
```
https://user-images.githubusercontent.com/7483112/216739787-3e481473-55f3-44e4-aa88-65baba49c0fa.mp4

Damn is that how many messages it was printing to chat? I know it consolidates them but 😬

Elli-Skala commented:
`refactor: Spiders and Geneticists now use the same ability to make webs rather than two virtually-identical ones, which should also perform just as it did before this change`

Does this mean spider webs no longer have the weird thing where they are faction based between who can walk freely through them? (Ie. Spiders struggle through someone's webs, and two people with webbing can't use each other's webs?)


Also does that mean genetics webs can make web walls now?

Jacquerel commented:
No and no, those seemed to be intentional features and thus still exist. 
I don’t have strong opinions about them but I can’t change the functionality of anything this month (much) or it would stop being a fix/refactor and start being a feature or balance pr.

Elli-Skala commented:
> No and no, those seemed to be intentional features and thus still exist. I don’t have strong opinions about them but I can’t change the functionality of anything this month (much) or it would stop being a fix/refactor and start being a feature or balance pr.

Then webbing being the strange only gene that still usable after being mutadoned remains, least for the next month

Jacquerel committed:
Merge branch 'master' into a-web-any-time

san7890 commented on code/modules/mob/living/basic/space_fauna/giant_spider/spider_abilities/hivemind.dm:
```diff
@@ -0,0 +1,70 @@
+	. = ..()
+	if(!.)
+		return
+	current_directive = tgui_input_text(owner, "Enter the new directive", "Create directive", "[current_directive]")
```
not sure how i feel about mutating current_directive directly like this in case of tgui_input_text schenanigans, have a local variable that we set it to after we nullcheck the results of the prompt. like
```txt
var/new_directive = tgui_input_text(...)
if(isnull(new_directive) || ...)
    return FALSE
current_directive = new_directive
```

juuuust to be super safe with handling. leaves more of a code footprint but i think it's worth it so we can "type" current_directive to a string as well rather than potentially null it.

san7890 commented on code/modules/mob/living/basic/space_fauna/giant_spider/spider_abilities/hivemind.dm:
```diff
@@ -0,0 +1,70 @@
+	background_icon_state = "bg_alien"
+	overlay_icon_state = "bg_alien_border"
+	check_flags = AB_CHECK_CONSCIOUS
+	/// Current directive to apply
+	var/current_directive
```
```suggestion
	/// Current directive to apply
	var/current_directive = ""
```

Type it to a string since it is one, do this in conjuction with the one where we don't mutate current_directive in the action's Trigger proc.

san7890 commented on code/modules/mob/living/basic/space_fauna/giant_spider/spider_abilities/hivemind.dm:
```diff
@@ -0,0 +1,70 @@
+/datum/action/command_spiders/proc/spider_command(mob/living/user, message)
+	if(!message)
+		return
+	var/my_message
+	my_message = span_spider("<b>Command from [user]:</b> [message]")
```
```suggestion
	var/my_message = span_spider("<b>Command from [user]:</b> [message]")
```
This should work, no need to be redundant.

san7890 commented on code/modules/mob/living/basic/space_fauna/giant_spider/spider_abilities/lay_eggs.dm:
```diff
@@ -0,0 +1,84 @@
+
+/datum/action/lay_eggs/IsAvailable(feedback = FALSE)
+	. = ..()
+	if(!.)
+		return FALSE
+	if(!isspider(owner))
+		return FALSE
```
```suggestion
	if(!. || !isspider(owner))
		return FALSE
```
why not do something like this?

Jacquerel replied:
I actually meant to remove that check and missed this one, it doesn't have a useful _functional_ reason to exist at this point and while it's only used by spiders right now there doesn't seem to be a reason to enforce it.

san7890 commented on code/modules/mob/living/basic/space_fauna/giant_spider/spider_abilities/wrap.dm:
```diff
@@ -0,0 +1,113 @@
+
+/datum/action/cooldown/wrap/IsAvailable(feedback = FALSE)
+	. = ..()
+	if(!.)
+		return FALSE
+	if(owner.incapacitated())
+		return FALSE
+	if(DOING_INTERACTION(owner, DOAFTER_SOURCE_SPIDER))
+		return FALSE
```
```suggestion
	if(!. || owner.incapacitated() || DOING_INTERACTION(owner, DOAFTER_SOURCE_SPIDER))
		return FALSE
```

san7890 commented on code/modules/mob_spawn/ghost_roles/spider_roles.dm:
```diff
@@ -102,11 +102,9 @@
+/obj/effect/mob_spawn/ghost_role/spider/special(mob/living/simple_animal/hostile/giant_spider/spawned_mob, mob/mob_possessor)
+	if (spawned_mob)
+		spawned_mob.directive = directive
 	. = ..()
```
Any specific reason why we set the directive before we call parent?

Jacquerel replied:
Yes, in parent it assigns the mind and then in the `login` proc for spiders it assigns the antag datum, which prints their objectives into the chat and pops up the antag panel.
We need to give them their objective _before_ that or it doesn't show it to them, which is a bug on live at the moment.
This is pretty fragile but I couldn't think of a way to make it less so.

san7890 replied:
Why can't we just `return ..()` at the end of the proc then?

Jacquerel replied:
You're right, the egg cleanup is old code but I don't know why I didn't just move all of it above the super call.

san7890 reviewed (changes requested):
I'm fearful of changing the default variable directly with tgui_input due to potential schenanigans but otherwise looks good beyond a few other things

Jacquerel committed:
Consolidates some checks
Removes pointless typecheck
Adds feedback to same place we use it on other abilities
Don't set stored variable directly to tgui input

Jacquerel committed:
Removes pointless check, nothing else assumes this could be null

Jacquerel committed:
Moves call to super to more sensible location.

Jacquerel committed:
again again

Jacquerel committed:
once more with feeling

san7890 commented on code/modules/mob/living/basic/space_fauna/giant_spider/spider_abilities/wrap.dm:
```diff
@@ -0,0 +1,113 @@
+		owner.balloon_alert(owner, "must be closer!")
+		return FALSE
+
+	if(!ismob(to_wrap) && !isobj(to_wrap))
+		return FALSE
+
+	if(to_wrap == owner)
+		return FALSE
```
```suggestion
	if(!ismovable(to_wrap) || (to_wrap == owner))
		return FALSE
```

Simplify the check (since we cast to `/atom/movable` anyways) as well as condense down the footprint of early return checks.

san7890 reviewed (changes requested):
Last thing, hopefully good to go after this.

Jacquerel committed:
Consolidate one more check
